rfc/system/0000-flyte-agent.md

```diff
@@ -0,0 +1,141 @@
+
+The Agent uses a long-poll pull-based gRPC request to both retreive and report the state of task executions. This effectively minimizes the latency of both operations as FlytePropeller, during initial task launch or status check, will immediately have access to an open connection from the appropriate Agent. Additionally, Agent initiated communiction enables fast reporting of task completion.
+
+**Flyte Agent Plugin**: The Flyte Agent Plugin is, quite obviously, a Flyte plugin which bridges Flytes native scheduling logic and Flyte Agent execution. By implementing this as a separate plugin, we can enable / disable easily and alternatively, set as the default for specific Flyte task types (ex. "python-task").
+
+The plugin contains a passive gRPC service to handle the pull-based connections from each Flyte Agent. As part of the gRPC request handling, the plugin maintains connection pools of agents based on either the agent ID, for retrieving the next task, or the task ID for reporting task status. Internally, this will be architected using a sequence of golang channels / callback functions to emulate a task queue.
+
+Durability of task executions is maintained using the existing `pluginStateBytes` construct within the FlyteWorkflow CRD. To begin, this may store the owning agent ID, the last successfuly sync (for liveness), and the current plugin phase. There are a number of challenges in providing exactly once execution semantics (if desired) during failures between gRPC requests and FlyteWorkflow CR persistence. We discuss a 2-phase commit lie protocol to combat these issues in TODO.
+
+### Protocol Deep Dive
+We purposfully introduced the components with a high-level, abstract description as many of the protocol intricasies compicate this view. However, it is important to explicitely outline implementation specifics.
+
+**Task Definition**: Currently, Flyte uses a simple `pyflyte execute ...` command within k8s Pods to describe the the task execution. For an initial solution, given the python entrypoint from PyO3 in Rust, this is a natural task definition. Future work may support more verbose task configurations with the potential for agents to have different execution environments beyond python.
+
+**Queue Topics**: The notion of a "queue topic" (also refered to as "topic") is used to link Flyte task definitions with Flyte Agent instances that are qualified to execute them. Within the Flyte Agent configuration, this can be as simple as including a `topic` configuration option. On the flytekit side, we can introduce an additional field into the task decorator, for example:
+
+`@task(topic="test-1")`
+
+It will be implicit that introduction of the `topic` field will make `resources`, `task_config`, and other fields moot. If this is not desired, we can introduce a new flytekit decorator, with something like `@fasttask` to note that this task should be executed on an agent
```

Comment:
i feel like it should be more transparent to the user... why not something like `@task(..., experimental_container_reuse=True)`?  That way we can just make it default one day?  or is the plan that this will never make it to production?